File: examples/benchmarks/3_scattering_kernel_consistency/compute_kernel.py
import ehtim as eh
import numpy as np
import pandas as pd
import timeit
import tqdm
import random

print("This script will benchmark eht-imaging's computation of scattering kernel.")

# number of uvsamples
ndata = 1000

# maximum uvd
uvdmax1 = 15e9  # 15 Gλ
uvdmax2 = 30e6  # 30 Mλ

# uv
uvd1 = np.linspace(0, uvdmax1, ndata)
uvd2 = np.linspace(0, uvdmax2, ndata)

# reference frequency
λcm1 = 0.087
λcm2 = 3.6

# scattering model
sm = eh.scattering.ScatteringModel(POS_ANG=0.0)

df = pd.DataFrame()
df["uvd1"] = uvd1
df["vismaj1"] = sm.Ensemble_Average_Kernel_Visibility(
    u=np.zeros(ndata), v=uvd1, wavelength_cm=λcm1
)
df["vismin1"] = sm.Ensemble_Average_Kernel_Visibility(
    u=uvd1, v=np.zeros(ndata), wavelength_cm=λcm1
)
df["uvd2"] = uvd2
df["vismaj2"] = sm.Ensemble_Average_Kernel_Visibility(
    u=np.zeros(ndata), v=uvd2, wavelength_cm=λcm2
)
df["vismin2"] = sm.Ensemble_Average_Kernel_Visibility(
    u=uvd2, v=np.zeros(ndata), wavelength_cm=λcm2
)
df.to_csv("eh_compute_kernel.csv")

File: examples/benchmarks/3_scattering_kernel_consistency/compute_kernel.jl
using Pkg
Pkg.activate(dirname(@__DIR__))
Pkg.instantiate()

using CSV
using BenchmarkTools
using DataFrames
using ScatteringOptics
using VLBISkyModels

@info "Computing scattering kernels with ScatteringOptics.jl"
# number of uvsamples
const ndata = 1000

# maximum uvd
const uvdmax1 = 15e9 # 15 Gλ
const uvdmax2 = 30e6 # 30 Mλ

# uv
const uvec1 = LinRange(0, uvdmax1, ndata)
const uvec2 = LinRange(0, uvdmax2, ndata)

# reference frequency
const λcm1 = 0.087
const λcm2 = 3.6

# creating the scattering models
const sm = ScatteringModel(; ϕpa_deg=0.0)

# a function to compute visibility
@inline visibilitymap_approx_min(sm, λcm, uvec) =
    map(u -> visibility_point_approx(sm, λcm, u, 0.0), uvec)
@inline visibilitymap_approx_maj(sm, λcm, vvec) =
    map(v -> visibility_point_approx(sm, λcm, 0.0, v), vvec)

# Benchmarking
vismaj1 = visibilitymap_approx_maj(sm, λcm1, uvec1)
vismin1 = visibilitymap_approx_min(sm, λcm1, uvec1)
vismaj2 = visibilitymap_approx_maj(sm, λcm2, uvec2)
vismin2 = visibilitymap_approx_min(sm, λcm2, uvec2)

df = DataFrame(;
    uvd1=uvec1,
    vismaj1=vismaj1,
    vismin1=vismin1,
    uvd2=uvec2,
    vismaj2=vismaj2,
    vismin2=vismin2,
)
CSV.write("so_compute_kernel.csv", df)